qualification/general/test_control.py

```diff
@@ -0,0 +1,206 @@
+        with check:
+            assert elapsed < 1.0
+        # Set next set of delays to be consistent with the old delay and delay_rate
+        delay += delay_rate * DELAY_INTERVAL
+        phase += phase_rate * DELAY_INTERVAL
+        delay = np.clip(delay, 0.0, MAX_DELAY)
+        phase = wrap_angle(phase)
+        delay_rate = rng.uniform(-MAX_DELAY_RATE, MAX_DELAY_RATE, n_inputs)
+        phase_rate = rng.uniform(-MAX_PHASE_RATE, MAX_PHASE_RATE, n_inputs)
+
+
+async def control_tacv_delays(rng: np.random.Generator, cbf: CBFRemoteControl, pdf_report: Reporter, name: str) -> None:
+    pcc = cbf.product_controller_client
+    src_stream = cbf.config["outputs"][name]["src_streams"][0]
+    n_inputs: int = len(cbf.config["outputs"][src_stream]["input_labels"]) // 2
+    async for _ in periodically(rng, BEAM_DELAY_INTERVAL):
+        delay = rng.uniform(-BEAM_MAX_DELAY, BEAM_MAX_DELAY, n_inputs)
+        phase = rng.uniform(-np.pi, np.pi, n_inputs)
+        coeff = [f"{d}:{p}" for d, p in zip(delay, phase)]
+        elapsed = await measure(pcc.request("beam-delays", name, *coeff))
+        pdf_report.detail(f"Set delays for {name} in {elapsed:.3f} s.")
+        with check:
+            assert elapsed < 1.0
+
+
+@pytest.fixture
+async def sensor_watcher(cbf: CBFRemoteControl) -> AsyncGenerator[aiokatcp.SensorWatcher, None]:
+    # aiokatcp doesn't currently handle adding watchers after the connection
+    # is already established; SensorWatcher is also somewhat expensive. So
+    # instead we create a separate connection for monitoring sensors.
+    secondary = aiokatcp.Client(*cbf.product_controller_endpoint)
+    sensor_watcher = aiokatcp.SensorWatcher(secondary)
+    secondary.add_sensor_watcher(sensor_watcher)
+    await sensor_watcher.synced.wait()
+    yield sensor_watcher
+    secondary.close()
+    await secondary.wait_closed()
+
+
+def check_timestamps(
+    name: str,
+    receiver: XBReceiver,
+    pdf_report: Reporter,
+    timestamps: list[int],
+):
+    assert timestamps, f"No {name} chunks received"
+    elapsed = (timestamps[-1] - timestamps[0]) / receiver.scale_factor_timestamp
+    pdf_report.detail(f"{name}: received data over {elapsed:.3f}s.")
+    with check:
+        assert elapsed >= 55.0, f"Less than 55s of data received for {name}"
+    # Trim the first 5 seconds, where we expect to see some losses
+    start = timestamps[0] + 5 * receiver.scale_factor_timestamp
+    start_pos = 0
+    while timestamps[start_pos] < start:
+        start_pos += 1
+    del timestamps[:start_pos]
+    expected = (timestamps[-1] - timestamps[0]) // receiver.timestamp_step + 1
+    missing = expected - len(timestamps)
+    pdf_report.detail(f"{name}: missed {missing} of {expected} chunks (ignoring first 5s).")
+    with check:
+        assert missing <= 1, f"{missing} of {expected} chunks missing for {name}"
+
+
+# TODO: once requirements spec is finalised, note which requirements
+# this corresponds to.
```

Comment:
In the old AQF, there was an "informal" section which ran tests for the engineers' sanity that everything was working, not specifically linked to a requirement. So we can have this if we want to.

Reply:
This will have linked requirements though, about the frequency with which delays need to be updated.